src/keys.rs

```diff
@@ -79,18 +202,20 @@ pub struct Diversifier([u8; 11]);
 /// This key is not suitable for use on its own in a wallet, as it cannot maintain
 /// accurate balance. You should use a [`FullViewingKey`] instead.
 #[derive(Debug)]
-pub struct IncomingViewingKey;
+pub struct IncomingViewingKey(pallas::Scalar);
 
 impl From<&FullViewingKey> for IncomingViewingKey {
-    fn from(_: &FullViewingKey) -> Self {
-        todo!()
+    fn from(fvk: &FullViewingKey) -> Self {
+        let ak = pallas::Point::from_bytes(&(&fvk.ak.0).into()).unwrap();
+        IncomingViewingKey(commit_ivk(&extract_p(&ak), &fvk.nk.0, &fvk.rivk))
     }
 }
 
 impl IncomingViewingKey {
     /// Returns the payment address for this key corresponding to the given diversifier.
-    pub fn address(&self, _: Diversifier) -> Address {
-        todo!()
+    pub fn address(&self, d: Diversifier) -> Address {
```

Comment:
Perhaps an `impl From<(IncomingViewingKey, Diversifier)> for Address`?

Reply:
I dislike overusing `From<(tuple)>` in APIs; the resulting `(a, b).into()` makes introspection of the code difficult without an IDE.

Reply:
I find it nice as the docs then generate all the `From`s for an `Address` or another type, all together, but

Reply:
Unfortunately I can't define this anymore, because `IncomingViewingKey::address` now returns `Option<Address>` to match the protocol spec, and both `From` and `Option` are foreign types.